File: examples/cpp/rename.cc
/*
 * Copyright (c) 2026 dingodb.com, Inc. All Rights Reserved
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// rename.cc — rename a file, verify via stat via the C API.

#include <fcntl.h>
#include <sys/stat.h>

#include <iostream>

#include "common.h"

int main() {
  uintptr_t h = make_client();

  DINGOFS_CHECK(dingofs_mount(h, MDS_ADDRS, FS_NAME, MOUNT_POINT), "dingofs_mount");

  // Setup: create directory + file
  DINGOFS_CHECK(dingofs_mkdir(h, "/demo_dir", 0755), "dingofs_mkdir");

  int fd = dingofs_open(h, "/demo_dir/old_name.txt",
                        O_WRONLY | O_CREAT | O_TRUNC, 0644);
  if (fd < 0) {
    std::cerr << "[FAIL] dingofs_open: " << strerror(-fd) << "\n";
    dingofs_umount(h);
    dingofs_delete(h);
    return 1;
  }

  // Record original inode
  struct stat orig {};
  dingofs_fstat(h, fd, &orig);
  std::cout << "original ino: " << orig.st_ino << "\n";

  DINGOFS_CHECK(dingofs_close(h, fd), "dingofs_close");

  // Rename
  DINGOFS_CHECK(
      dingofs_rename(h, "/demo_dir/old_name.txt", "/demo_dir/new_name.txt"),
      "dingofs_rename");

  // Verify via stat on new name — inode must be the same
  struct stat renamed {};
  DINGOFS_CHECK(dingofs_stat(h, "/demo_dir/new_name.txt", &renamed),
                "dingofs_stat(new_name)");
  std::cout << "renamed  ino: " << renamed.st_ino << "  (was " << orig.st_ino
            << ")\n";

  // Cleanup
  DINGOFS_CHECK(dingofs_unlink(h, "/demo_dir/new_name.txt"), "dingofs_unlink");
  DINGOFS_CHECK(dingofs_rmdir(h, "/demo_dir"), "dingofs_rmdir");

  DINGOFS_CHECK(dingofs_umount(h), "dingofs_umount");
  dingofs_delete(h);
  return 0;
}

File: examples/python/rename.py
# Copyright (c) 2026 dingodb.com, Inc. All Rights Reserved
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""rename.py — rename a file and verify via stat."""

import sys
from dingofs import DingofsError
from common import make_client

client = make_client()
try:
    client.mkdir("/demo_dir", 0o755)
    print("[ OK ] MkDir")

    with client.open("/demo_dir/old_name.txt", "wb") as f:
        f.write(b"data")
    old_attr = client.stat("/demo_dir/old_name.txt")
    print(f"[ OK ] Create  ino={old_attr.ino}")

    # Rename
    client.rename("/demo_dir/old_name.txt", "/demo_dir/new_name.txt")
    print("[ OK ] Rename")

    # Verify: old path is gone, new path resolves to the same inode
    renamed = client.stat("/demo_dir/new_name.txt")
    print(f"[ OK ] Stat    ino={renamed.ino}  (was {old_attr.ino})")

    client.unlink("/demo_dir/new_name.txt")
    client.rmdir("/demo_dir")
    print("[ OK ] Cleanup")

except DingofsError as e:
    print(f"[FAIL] {e}", file=sys.stderr)
    sys.exit(1)
finally:
    client.stop()
    print("[ OK ] Stop")
